Mods/SML/Source/SML/Private/Patching/BlueprintHookManager.cpp

```diff
@@ -35,44 +35,35 @@ void DebugDumpFunctionScriptCode(UFunction* Function, int32 HookOffset, const FS
 }
 #endif
 
-void UBlueprintHookManager::InstallBlueprintHook(UFunction* Function, int32 HookOffset) {
+void UBlueprintHookManager::InstallBlueprintHook(UFunction* Function, const int32 RequestedHookOffset, const int32 ResolvedHookOffset) {
 	TArray<uint8>& OriginalCode = Function->Script;
-	checkf(OriginalCode.Num() > HookOffset, TEXT("Invalid hook: HookOffset > Script.Num()"));
+	fgcheckf(OriginalCode.Num() > ResolvedHookOffset, TEXT("Invalid hook: Resolved HookOffset > Script.Num()"));
 
 #if DEBUG_BLUEPRINT_HOOKING
-	DebugDumpFunctionScriptCode(Function, HookOffset, TEXT("BeforeHook"));
+	DebugDumpFunctionScriptCode(Function, RequestedHookOffset, ResolvedHookOffset, TEXT("BeforeHook"));
 #endif
-	
-	//Minimum amount of bytes required to insert unconditional jump with code offset
-	const int32 MinBytesRequired = 1 + sizeof(CodeSkipSizeType);
 
+	// Will throw an error if the resolved hook offset is not at properly-aligned, parseable opcode
 	FSMLKismetBytecodeDisassembler Disassembler;
-	int32 BytesAvailable = 0;
-	
-	//Walk over statements until we collect enough bytes for a replacement
-	//(or until we consumed all statements in the function's code)
-	while (BytesAvailable < MinBytesRequired && (HookOffset + BytesAvailable) < OriginalCode.Num()) {
-		const int32 CurrentStatementIndex = HookOffset + BytesAvailable;
-		int32 OutStatementLength;
-		
-		const bool bValid = Disassembler.GetStatementLength(Function, CurrentStatementIndex, OutStatementLength);
-		checkf(bValid, TEXT("Provided hook offset is not a valid statement index: %d"), HookOffset);
-		BytesAvailable += OutStatementLength;
+	Disassembler.SerializeStatement(Function, ResolvedHookOffset);
```

Comment:
I'm not sure if I understand why this is needed. Wouldn't `SerializeFunction` also assert if there's a problem?

Reply:
No, serializing the whole function doesn't care about the resolved hook offset and would succeed, at least for the first hook insertion.  This functions as an assertion that the resolved hook offset was resolved sanely and hooking won't obviously corrupt the function.  Without this assert, the failure mode of an error would be a crash when the hooked function was invoked.

BUT... this is admittedly relying on the fgcheckf in the disassembler - I don't love that but I'm not sure how to best do error handling here.  If the resolved offset is wrong, it WILL cause a crash in the disassembler.  If it threw an exception, I would catch it here and fail with a better error message.  If there's a more idiomatic way to handle that scenario, let me know (this is also part of why I updated the error text on that fgcheckf, btw).

Reply:
I can't think of a more elegant approach right now, so let's leave it be.

Ack